### ISSUE TITLE
Enums are generated incorrectly

### ISSUE THREAD

**Issue by deser** [dotansimha/graphql-code-generator#3132]
**Describe the bug**
Graphql Enums have keys and values. Seems graphql-codegen uses only keys for generaring enums in typescript.

See example below: 
```ts
// Such GQL enum 
new GraphQLEnumType({
	name: 'FvFormFillerServiceControlTypes',
	values: {
		Text1: {
			value: 'lala'
		}
	}
})

// will be converted into such typescript enum:

export enum GQLFvFormFillerServiceControlTypes {
  Text1 = 'Text1',
}

// Expected typescript enum is:
export enum GQLFvFormFillerServiceControlTypes {
  Text1 = 'lala',
}
```

**Comment by BigsonLvrocha** [comment]
in this case you will have to map your enum values in your codegen.yml file, field 
```yml
generates:
  schema.ts:
    plugins:
      - typescript
    config:
      enumValues:
        FvFormFillerServiceControlTypes:
          Text1: lala
```

**Comment by deser** [comment]
This is obviously incorrect behavior. GraphQL has in its enum key and value. Typescript has in its enum key and value. It seems logically correct to map key to key and value to value. Codegen should generate code rather than make me insert manually such enums :)
I believe I can even create pull for that:)

**Comment by BigsonLvrocha** [comment]
I didn't read the code, but my guess would be that they work with the resulting generated .graphql file or just consider not using the defined values to cover for frontend users, who doesn't have information about internal enum mapping in the server and will be consuming the graphql only with the plain values. 
Maybe there should be an option to chose between using the internal values in enum or not

**Comment by deser** [comment]
>Maybe there should be an option to chose between using the internal values in enum or not
As an non-breaking feature - definitely

**Comment by ardatan** [comment]
@deser You can use `enumValues` for internal mapping.
https://graphql-code-generator.com/docs/plugins/typescript#enumvalues-enumvaluesmap

**Comment by deser** [comment]
This was suggested above and I described why this doesn't match :)

**Comment by ardatan** [comment]
Currently GraphQL Codegen uses definitions only in the visitor implementations of all plugins. It doesn't respect the actual `GraphQLNamedType`s like `GraphQLEnumType` as in your example. It introspects the schema and parses again to make sure it has a valid `astNode` to make the visitors work correctly.
I pushed some changes on GraphQL Toolkit to keep actual `GraphQLSchema` object by fixing its `astNode` to make it work properly with the visitors instead of printing and reparsing.
https://github.com/ardatan/graphql-toolkit/commit/4d94736a02498d267724e0e461dddcb719e84247
I also did some changes on TypeScript visitor to use existing enum values if there is any defined in `GraphQLEnumType`.
https://github.com/dotansimha/graphql-code-generator/pull/3147
When all those come together, we will directly use your `GraphQLSchema` object with your `GraphQLNamedType`s but we will fix it automatically inside GraphQL Toolkit if there is any problem with this `GraphQLSchema` object.
It might be better to call it as an enhancement than a bug because every single type in `typescript` package is generated to be used as typings not actual code.
I'll keep you posted when the PR is ready to test :)

**Comment by BigsonLvrocha** [comment]
> every single type in typescript package is generated to be used as typings not actual code.
that may be the tricky part, since enums will be compiled to actual objects which may be used for for code

**Comment by ardatan** [comment]
@BigsonLvrocha Could you try with the following version?
`1.9.2-alpha-1455536b.55+1455536b`

**Comment by deser** [comment]
@ardatan ,
Firstly - thanks a lot :)
Secondly - yes, I'll try 1.9.2-alpha-1455536b.55+1455536b. Will respond soon

**Comment by deser** [comment]
Updated:
```
"@graphql-codegen/cli": "1.9.2-alpha-1455536b.55",
"@graphql-codegen/typescript": "1.9.2-alpha-1455536b.55",
```

Seeing errors:
```
 Unable to download schema from remote: Unknown operation named "introspection".
        Error: Unable to download schema from remote: Unknown operation named "introspection".
```

**Comment by ardatan** [comment]
@deser I fixed the introspection query in URL loader but I don't think it is possible to get enum values from the schema's introspection because it is not part of schema definition/SDL. Could you try with the actual code file that exports schema?

**Comment by deser** [comment]
Last time I tried I failed to do it. We define schema in old school way (literally constructing `new GraphQLObjectType` ) and writing script which requires that schema as a module and generates graphql types failed. Therefore the only way I found working is introspection schema :)

Let me try again maybe something has changed from the last time. 
Should I install 1.9.2-alpha-1455536b.55 again?

**Comment by ardatan** [comment]
You can try it with `1.9.2-alpha-74dff252.63+74dff252`. I added tests for your case.

**Comment by deser** [comment]
ok thx

**Comment by deser** [comment]
Doesnt work with introspection schema. Will try to work with schema directly

**Comment by ardatan** [comment]
> @deser I fixed the introspection query in URL loader but I don't think it is possible to get enum values from the schema's introspection because it is not part of schema definition/SDL. Could you try with the actual code file that exports schema?

@deser It will never work with introspection and SDL because enum values are not part of the schema definition. You should try the exported schema as I said above.

**Comment by deser** [comment]
Sure, I'm trying that. Thank you

**Comment by deser** [comment]
It works. Thanks a lot!

**Comment by deser** [comment]
Found a possible issue:
```
config:
            declarationKind: 'interface'
            typesPrefix: GQL
            skipTypename: true
            addUnderscoreToArgsType: true
            maybeValue: T
```

With this config it generates incorrect types:
![ngen_screen_designer  D__work_ngen_screen_designer (32)](https://user-images.githubusercontent.com/3010698/71474695-b542ae80-27ed-11ea-85d5-b838d14c10fc.png)

**Comment by deser** [comment]
Found a possible issue:
```
config:
            declarationKind: 'interface'
            typesPrefix: GQL
            skipTypename: true
            addUnderscoreToArgsType: true
            maybeValue: T
```

With this config it generates incorrect types:
![ngen_screen_designer  D__work_ngen_screen_designer (32)](https://user-images.githubusercontent.com/3010698/71474695-b542ae80-27ed-11ea-85d5-b838d14c10fc.png)

**Comment by deser** [comment]
Previously it was ok:
![image](https://user-images.githubusercontent.com/3010698/71474903-c8a24980-27ee-11ea-881b-2f5fa89295d3.png)

And adding `enumPrefix: GQL` doesn't help. Seems bug

**Comment by dotansimha** [comment]
I think it should be fixed in `v1.10.0`.
@deser let me know if it needs to be reopened and fixed again.

**Comment by deser** [comment]
Ok thx a lot!
